pkg/config/expand.go

```diff
@@ -339,6 +339,9 @@ func validateModuleSettingReference(bp Blueprint, mod Module, r Reference) error
 	}
 
 	if err := validateModuleReference(bp, mod, r.Module); err != nil {
+		if _, ok := err.(InvalidModuleError); ok && bp.Vars.Has(r.Name) {
```

Comment:
* IIUC. `err` will be of type `HintError` instead of `InvalidModuleError` if you implement it in `bp.Module()`, so you need to use `errors.Is` instead of casting.
*  The chance of having the same name for global vars and module outputs is hight (> 50%). It may not be reasonable to always suggest "vars" here. 
* Avoid double-wrapping.

I would advice to not do hinting in `Blueprint.Module`.

Reply:
> * IIUC. `err` will be of type `HintError` instead of `InvalidModuleError` if you implement it in `bp.Module()`, so you need to use `errors.Is` instead of casting.

No, I don't think that's what it does.  If there isn't a close enough match, an InvalidModuleError is returned.  Either way I can switch out the type check for an error.Is function.

> * The chance of having the same name for global vars and module outputs is hight (> 50%). It may not be reasonable to always suggest "vars" here.

I don't think I always suggest it.  The logic is that if the error is an InvalidModuleError, that means no other module name was close enough to warrant a HintError and so nothing has been suggested yet.  As a result then we check if the reference name is within the pool of global variables, and if it is, then it is hinted.  Otherwise, nothing is suggested.

> * Avoid double-wrapping.

Is there no value in maintaining the error type from below?  We might want to check what kind of error is below the Hint, or just remove the HintError type altogether and just have hints as optional portions of the InvalidModuleError.  Alternatively, we could make HintError as part of the InvalidModuleError for inheritance, that way other things we want to give suggestions on could follow suit.

> 
> I would advice to not do hinting in `Blueprint.Module`.

Is there a reason why, given my explanations above?

Reply:
Done.